plugins/out_s3/s3.c

```diff
@@ -1393,91 +1603,140 @@ static void cb_s3_flush(const void *data, size_t bytes,
-        if (timeout_check == FLB_FALSE) {
-            /* add data to local buffer */
-            ret = s3_store_buffer_put(ctx, chunk,
-                                      tag, tag_len, json, (size_t) len);
-            if (s3_plugin_under_test() == FLB_TRUE) {
-                goto send_data;
-            }
-            flb_sds_destroy(json);
+    /* File is ready for upload */
+    if ((upload_file && upload_file->size + chunk_size > ctx->upload_chunk_size) ||
+        (m_upload_file && m_upload_file->bytes + chunk_size > ctx->file_size) ||
+        upload_timeout_check == FLB_TRUE) {
+        /* Use upload queue */
+        if (ctx->preserve_data_ordering == FLB_TRUE) {
+            /* Buffer chunk in filesystem */
+            ret = s3_store_buffer_put(ctx, upload_file, tag, tag_len, chunk,
+                                      (size_t) chunk_size);
+            flb_sds_destroy(chunk);
             if (ret < 0) {
+                flb_plg_warn(ctx->ins, "Could not buffer chunk. Data order preservation "
+                             "will be compromised");
                 FLB_OUTPUT_RETURN(FLB_RETRY);
             }
+
+            /* Chunk file is ready, lock to prevent further changes */
+            s3_store_file_lock(upload_file);
+
+            /* Add chunk file to upload queue */
+            ret = add_to_queue(ctx, upload_file, m_upload_file, tag, tag_len);
+            if (ret < 0) {
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
+
+            /* Go through upload queue and return error if something went wrong */
+            s3_upload_queue(config, ctx);
+            if (ctx->upload_queue_success == FLB_FALSE) {
+                ctx->upload_queue_success = FLB_TRUE;
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
             FLB_OUTPUT_RETURN(FLB_OK);
         }
-    }
+        /* Upload each file as it comes */
+        else {
+            /* Prepare output buffer to upload */
+            ret = construct_request_buffer(ctx, chunk, upload_file, &buffer,
+                                           &buffer_size);
+            if (ret < 0) {
+                flb_plg_error(ctx->ins, "Could not construct request buffer for %s",
+                              upload_file->file_path);
+                FLB_OUTPUT_RETURN(FLB_ERROR);
+            }
 
-send_data:
-    ret = construct_request_buffer(ctx, json, chunk, &buffer, &buffer_size);
-    flb_sds_destroy(json);
-    if (ret < 0) {
-        flb_plg_error(ctx->ins, "Could not construct request buffer for %s",
-                      chunk->file_path);
-        FLB_OUTPUT_RETURN(FLB_RETRY);
-    }
+            /* Send upload request with buffer output */
+            ret = upload_data(ctx, upload_file, m_upload_file, buffer, buffer_size,
+                              tag, tag_len);
+            flb_free(buffer);
 
-    ret = upload_data(ctx, chunk, m_upload, buffer, buffer_size, tag, tag_len);
-    flb_free(buffer);
+            FLB_OUTPUT_RETURN(ret);
```

Comment:
shouldn't the return value be one of the FLB_ERROR, FLB_OK, FLB_RETRY? This is confusing.

Reply:
Yes, the output of `upload_data` will be one of those three. We want to output whatever `upload_data` outputs!